### PR TITLE
ms-gsl: Use validate method

### PR DESCRIPTION
**Pull request by jwillikers**
Specify library name and version:  **ms-gsl/***

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
Could you add a basic_layout please? There are known side effects in some generators (like PkgConfigDeps) when layout is missing.

**Comment by jwillikers** [comment]
> Could you add a basic_layout please? There are known side effects in some generators (like PkgConfigDeps) when layout is missing.

@SpaceIm No problem.

File: recipes/ms-gsl/all/conanfile.py [hunk 1]
```diff
@@ -37,37 +39,44 @@ def _contract_map(self):
     def _compilers_minimum_version(self):
         return {
             "gcc": "5",
-            "Visual Studio": "14",
             "clang": "3.4",
             "apple-clang": "3.4",
         }
 
+    def validate(self):
+        check_min_cppstd(self, 14)
```
Comment (SpaceIm) [inline recipes/ms-gsl/all/conanfile.py:47]:
```suggestion
        if self.settings.compiler.cppstd:
            check_min_cppstd(self, 14)
```

File: recipes/ms-gsl/all/conanfile.py [hunk 2]
```diff
@@ -37,37 +39,44 @@ def _contract_map(self):
     def _compilers_minimum_version(self):
         return {
             "gcc": "5",
-            "Visual Studio": "14",
             "clang": "3.4",
             "apple-clang": "3.4",
         }
 
+    def validate(self):
```
Comment (SpaceIm) [inline recipes/ms-gsl/all/conanfile.py:46]: validate() after package_id() please, to understand the flow (package id can remove things in self.info, and it must be taken into account validate()).

**Comment by SpaceIm** [comment]
@jgsogo all configurations of https://github.com/conan-io/conan-center-index/pull/12453#issuecomment-1222645338 are INVALID_CONFIGURATION, but bot didn't complain.

File: recipes/ms-gsl/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py]: Done.

File: recipes/ms-gsl/all/conanfile.py (hunk 2 above)
Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py:49]: Done.

File: recipes/ms-gsl/all/conanfile.py [hunk 3]
```diff
@@ -43,8 +43,13 @@ def _compilers_minimum_version(self):
             "apple-clang": "3.4",
         }
 
+    def package_id(self):
```
Comment (SpaceIm) [inline recipes/ms-gsl/all/conanfile.py:46]: and config_options is called before package_id also, this is why I said to move validate() ;)

Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py:50]: Done.

**Comment by jgsogo** [comment]
> @jgsogo all configurations of [#12453 (comment)](https://github.com/conan-io/conan-center-index/pull/12453#issuecomment-1222645338) are INVALID_CONFIGURATION, but bot didn't complain.

We decided long ago that merging recipes that didn't generate any binaries was ok. It was related to packages that required certain linux-headers that were not the ones _installed_ in the CI.

**Review by uilianries — APPROVED** [review]

**Comment by SpaceIm** [comment]
> > @jgsogo all configurations of [#12453 (comment)](https://github.com/conan-io/conan-center-index/pull/12453#issuecomment-1222645338) are INVALID_CONFIGURATION, but bot didn't complain.
> 
> We decided long ago that merging recipes that didn't generate any binaries was ok. It was related to packages that required certain linux-headers that were not the ones _installed_ in the CI.

Would be nice to show a warning somewhere, because for most recipes, it means that there is an issue.

**Comment by jgsogo** [comment]
Yes, instead of 
image

maybe a different message: `"Zero packages generated"`

**Review by jgsogo — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**